File: backend/app/schemas/documents.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.enums import MatchDecision, ProcessingStatus


class DocumentUploadResponse(BaseModel):
    document_id: int
    filename: str
    status: ProcessingStatus


class DocumentStatusResponse(BaseModel):
    document_id: int
    filename: str
    status: ProcessingStatus
    processing_started_at: datetime | None
    processing_completed_at: datetime | None
    error_message: str | None
    extracted_text_length: int | None


class DocumentProcessingResponse(BaseModel):
    document_id: int
    status: ProcessingStatus
    entities_found: int
    relationships_found: int
    evidence_created: int
    review_required: int
    graph_sync_status: str


class ExtractionResponse(BaseModel):
    id: int
    kind: str
    entity_type: str
    text: str
    normalized_value: str
    confidence: float
    page: int | None
    source_reference: str | None
    source_context: str | None
    review_status: str


class RelationshipExtractionResponse(BaseModel):
    id: int
    relationship_type: str
    source_entity: str
    target_entity: str
    confidence: float
    page: int | None
    source_reference: str | None
    source_text: str
    review_status: str


class DocumentExtractionsResponse(BaseModel):
    document_id: int
    status: ProcessingStatus
    entities: list[ExtractionResponse]
    relationships: list[RelationshipExtractionResponse]
    evidence: list[dict[str, Any]] = Field(default_factory=list)


class DocumentDetailResponse(DocumentStatusResponse):
    case_id: int
    document_type: str
    mime_type: str | None
    file_size_bytes: int | None
    upload_timestamp: datetime
    extraction_summary: dict[str, int]


class MatchReviewRequest(BaseModel):
    decision: MatchDecision


class ExtractionReviewRequest(BaseModel):
    decision: MatchDecision


class EntityMatchResponse(BaseModel):
    id: int
    extraction_id: int
    case_id: int
    candidate_entity_type: str
    candidate_entity_id: int | None
    candidate_label: str | None
    match_type: str
    confidence: float
    signals: dict[str, Any] | None
    status: str

File: frontend/src/types/documents.ts
export type ProcessingStatus = "PENDING" | "PROCESSING" | "COMPLETED" | "FAILED";

export type DocumentUploadResult = {
  document_id: number;
  filename: string;
  status: ProcessingStatus;
};

export type DocumentProcessingResult = {
  document_id: number;
  status: ProcessingStatus;
  entities_found: number;
  relationships_found: number;
  evidence_created: number;
  review_required: number;
  graph_sync_status: string;
};

export type DocumentStatus = {
  document_id: number;
  filename: string;
  status: ProcessingStatus;
  processing_started_at: string | null;
  processing_completed_at: string | null;
  error_message: string | null;
  extracted_text_length: number | null;
};

export type DocumentDetail = DocumentStatus & {
  case_id: number;
  document_type: string;
  mime_type: string | null;
  file_size_bytes: number | null;
  upload_timestamp: string;
  extraction_summary: {
    entities: number;
    relationships: number;
    evidence: number;
    review_required: number;
  };
};

export type Extraction = {
  id: number;
  kind: string;
  entity_type: string;
  text: string;
  normalized_value: string;
  confidence: number;
  page: number | null;
  source_reference: string | null;
  source_context: string | null;
  review_status: string;
};

export type RelationshipExtraction = {
  id: number;
  relationship_type: string;
  source_entity: string;
  target_entity: string;
  confidence: number;
  page: number | null;
  source_reference: string | null;
  source_text: string;
  review_status: string;
};

export type DocumentExtractions = {
  document_id: number;
  status: ProcessingStatus;
  entities: Extraction[];
  relationships: RelationshipExtraction[];
  evidence: Array<Record<string, unknown>>;
};

export type EntityMatch = {
  id: number;
  extraction_id: number;
  case_id: number;
  candidate_entity_type: string;
  candidate_entity_id: number | null;
  candidate_label: string | null;
  match_type: string;
  confidence: number;
  signals: Record<string, unknown> | null;
  status: string;
};

export type ReviewDecision = "ACCEPT" | "REJECT" | "DEFER";
